lib/ecto/migration/runner.ex

```diff
@@ -23,7 +23,9 @@ defmodule Ecto.Migration.Runner do
     metadata(runner, opts)
 
     log(level, "== Running #{version} #{inspect module}.#{operation}/0 #{direction}")
-    {time1, _} = :timer.tc(module, operation, [])
+    {time1, _} = :timer.tc(fn ->
+      perform_operation(module, operation)
```

Comment:
I would prefer to move this to `migrator` module, as that's the one that starts the transaction OR we move the logic that starts the transaction here. Otherwise the logic can easily get out of sync. For example, we should also check for `repo.__adapter__.supports_dll_transaction?` below, but we don't. :)

Reply:
Alright, so I've moved the calls into a sort of naive place, and they result in the error I mentioned. I don't see a great way around resolving that without moving the call to the callbacks into the `attempt/6` code. However, I'm unsure if it is correct to move the transaction start/end into that same place. Where do you think would be best?

Reply:
@zachdaniel On one hand, we don't need to make `execute` available in those callbacks. But that means we should probably pass the `repo` as argument.

Another option is to move where they were and use `repo.in_transaction?` to check if you are inside a transaction before running the callbacks?

Reply:
Hmm....Yeah, let’s just check if they are in a transaction with the repo :) good idea.

Reply:
@josevalim so checking if the repo is in a transaction presents a slight issue, which is that adapters raise a `RuntimeError: not implemented` when they don't implement `in_transaction?`. The only way I can think to get around that is to do (maybe with better matching):

```
in_transaction? =
  try do
    repo.in_transaction?()
  rescue
   RuntimeError -> false
  end
```

Reply:
That actually causes problems for tests, unfortunately. Without the test adapter being able to determine if its in a transaction, I can't really test this behavior without mocking/stubbing that functionality out somehow. Unless we maintain some corresponding kind of state in the process dictionary, but I have a feeling that could have some interesting repercussions.